compiler/luci/import/src/ImportMetadata.cpp

```diff
@@ -0,0 +1,161 @@
+{
+  std::map<uint32_t, std::set<uint32_t>> node_source_ids_map;
+
+  uint32_t idx = 0;
+  while (idx < buffer.size())
+  {
+    uint32_t id = 0;
+    id += (buffer.at(idx + 0) << 0 * 8);
+    id += (buffer.at(idx + 1) << 1 * 8);
+    id += (buffer.at(idx + 2) << 2 * 8);
+    id += (buffer.at(idx + 3) << 3 * 8);
+    idx += 4;
+
+    uint32_t node_num = 0;
+    node_num += (buffer.at(idx + 0) << 0 * 8);
+    node_num += (buffer.at(idx + 1) << 1 * 8);
+    node_num += (buffer.at(idx + 2) << 2 * 8);
+    node_num += (buffer.at(idx + 3) << 3 * 8);
+    idx += 4;
+
+    std::set<uint32_t> source_ids;
+    for (uint32_t j = 0; j < node_num; ++j)
+    {
+      uint32_t origin = 0;
+      origin += (buffer.at(idx + 4 * j + 0) << 0 * 8);
+      origin += (buffer.at(idx + 4 * j + 1) << 1 * 8);
+      origin += (buffer.at(idx + 4 * j + 2) << 2 * 8);
+      origin += (buffer.at(idx + 4 * j + 3) << 3 * 8);
+
+      source_ids.insert(origin);
+    }
+    idx += (node_num * 4);
+
+    if (node_source_ids_map.insert({id, source_ids}).second == false)
+      throw std::runtime_error("Duplicated origin ID");
+  }
+  assert(idx == buffer.size());
+
+  return node_source_ids_map;
+}
+
+/**
+ * 'source_table'
+ * - [ id : uint32_t ][ length : uint32_t ][ data : 'length' Bytes ]
+ */
+std::map<uint32_t, std::string> metadata_with_source_table(const std::vector<uint8_t> &buffer)
+{
+  std::map<uint32_t, std::string> source_id_name_map;
+
+  uint32_t idx = 0;
+  while (idx < buffer.size())
+  {
+    uint32_t id = 0;
+    id += (buffer.at(idx + 0) << 0 * 8);
+    id += (buffer.at(idx + 1) << 1 * 8);
+    id += (buffer.at(idx + 2) << 2 * 8);
+    id += (buffer.at(idx + 3) << 3 * 8);
+    idx += 4;
+
+    uint32_t length = 0;
+    length += (buffer.at(idx + 0) << 0 * 8);
+    length += (buffer.at(idx + 1) << 1 * 8);
+    length += (buffer.at(idx + 2) << 2 * 8);
+    length += (buffer.at(idx + 3) << 3 * 8);
+    idx += 4;
+
+    std::string origin_name;
+    for (uint32_t j = 0; j < length; ++j)
+      origin_name += (char)buffer.at(idx + j);
+    idx += length;
+
+    if (source_id_name_map.insert({id, origin_name}).second == false)
+      throw std::runtime_error("Duplicated origin ID");
+  }
+  assert(idx == buffer.size());
+
+  return source_id_name_map;
+}
+
+} // namespace
+
+namespace luci
+{
+
+ImportMetadata::ImportMetadata(const luci::CircleReader &reader)
+{
+  std::map<uint32_t, std::set<uint32_t>> node_source_ids_map;
+  std::map<uint32_t, std::string> source_id_name_map;
+
+  const auto &metadata = reader.metadata();
+
+  for (uint32_t i = 0; i < metadata.size(); ++i)
+  {
+    const circle::MetadataT &meta = *metadata[i];
+
+    assert(meta.buffer < reader.buffers().size());
+    const std::vector<uint8_t> &buffer = reader.buffers()[meta.buffer]->data;
+
+    // TODO Handle this by enum
+    if (meta.name.compare("op_table") == 0)
+    {
+      node_source_ids_map = metadata_with_op_table(buffer);
+    }
+    else if (meta.name.compare("source_table") == 0)
```

Comment:
(Originally written by @seanshpark)

are these to `op_table` and `source_table` are used for key to find origin data?
- is this format our own or from schema?

if this is from schema, we may consider conflict with tflite file improt.
- It would be better to use one key for origin data instead of two items
- to be safe, it would be better to use some prefix for our keys... this is the first items so I think this is very important
- I suggest to use `ONE_NodeOrigin` as the key

You wrote `// TODO Handle this by enum`... can you explain how you mean by `enum`?

Reply:
> are these to `op_table` and `source_table` are used for key to find origin data?
> - is this format our own or from schema?

Yes, this is our own format.
Schema only provides a placeholder for storing the data. (`table Metadata`)

> if this is from schema, we may consider conflict with tflite file improt.
> - It would be better to use one key for origin data instead of two items
> - to be safe, it would be better to use some prefix for our keys... this is the first items so I think this is very important
> - I suggest to use `ONE_NodeOrigin` as the key

For now, tflite only produces `min_runtime_version` metadata so there is no problem.
However, to prevent future conflicts, using some prefix for our metadata seems good.
( `ONE_` looks good to me)
Nevertheless, I think integrating the two table as one is not so good.

- Two table include different information.
- If one table include huge amount of data, it is hard to manage them.
- Backend is also using two kinds of table. There is no obligation to follow exactly same format, but similar format looks better I think.

> You wrote `// TODO Handle this by enum`... can you explain how you mean by `enum`?

This is minor thing :D
For now, I wrote a function like `metadata_with_op_table(buffer)`.
I thought which is better, `metadata_with_op_table(buffer)` or `metadata<OPTABLE>(buffer)`
`OPTABLE` is enum :)
`TODO` in here means this code is not fixed and I am thinking...!